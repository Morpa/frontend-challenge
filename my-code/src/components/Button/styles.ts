import styled, { css, DefaultTheme } from 'styled-components'

import { ButtonProps } from '.'

export type WrapperProps = {
  hasIcon: boolean
} & Pick<ButtonProps, 'favourite'>

const wrapperModifiers = {
  withIcon: (theme: DefaultTheme) => css`
    svg {
      width: 1.5rem;

      & + span {
        margin-left: ${theme.spacings.xxsmall};
      }
    }
  `,
  favourited: (theme: DefaultTheme) => css`
    background: ${theme.colors.red};

    svg {
      color: ${theme.colors.white};
      width: 1.5rem;

      & + span {
        margin-left: ${theme.spacings.xxsmall};
      }
    }
  `,
  addFavourite: (theme: DefaultTheme) => css`
    color: ${theme.colors.lightGrey};
    background: none;
    border: 1px solid ${theme.colors.lightGrey};
    border-radius: ${theme.border.radius};

    svg {
      width: 1.5rem;

      & + span {
        margin-left: ${theme.spacings.xxsmall};
      }
    }
  `
}

export const Wrapper = styled.button<WrapperProps>`
  ${({ theme, hasIcon, favourite }) => css`
    display: inline-flex;
    align-items: center;
    justify-content: center;
    color: ${theme.colors.white};
    border: 0;
    cursor: pointer;
    border-radius: ${theme.border.radius};
    padding: ${theme.spacings.xxsmall};
    height: 4rem;
    font-size: ${theme.font.sizes.medium};
    font-weight: ${theme.font.bold};

    ${!!hasIcon && wrapperModifiers.withIcon(theme)};
    ${!favourite && wrapperModifiers.addFavourite(theme)};
    ${!!favourite && wrapperModifiers.favourited(theme)};
  `}
`
