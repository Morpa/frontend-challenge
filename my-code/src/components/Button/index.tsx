import { ButtonHTMLAttributes } from 'react'
import * as S from './styles'

type ButtonTypes = ButtonHTMLAttributes<HTMLButtonElement>

export type ButtonProps = {
  icon?: React.ReactNode
  favourite: boolean
} & ButtonTypes

const Button = ({ children, icon, favourite, ...props }: ButtonProps) => (
  <S.Wrapper hasIcon={!!icon} {...props} favourite={favourite}>
    {!!icon && icon}
    {!!children && <span>{children}</span>}
  </S.Wrapper>
)

export default Button
