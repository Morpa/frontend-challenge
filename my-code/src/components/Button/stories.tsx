import { Story, Meta } from '@storybook/react/types-6-0'
import { Favorite, FavoriteBorder } from '@styled-icons/material-outlined'

import Button, { ButtonProps } from '.'

export default {
  title: 'Button',
  component: Button,
  argTypes: {
    children: {
      type: 'string'
    },
    icon: {
      type: ''
    }
  }
} as Meta

export const Default: Story<ButtonProps> = (args) => <Button {...args} />

Default.args = {
  children: 'Add to favourites',
  icon: <FavoriteBorder />,
  favourite: false
}

export const FavoritedButton: Story<ButtonProps> = (args) => (
  <Button {...args} />
)

FavoritedButton.args = {
  children: 'Added',
  icon: <Favorite />,
  favourite: true
}
