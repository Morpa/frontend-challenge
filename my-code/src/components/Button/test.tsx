import { screen } from '@testing-library/react'
import { renderWithTheme } from 'utils/tests/helpers'
import { Favorite, FavoriteBorder } from '@styled-icons/material-outlined'

import Button from '.'

describe('<Button />', () => {
  it('should render the favorited button', () => {
    const { container } = renderWithTheme(
      <Button icon={<Favorite data-testid="icon" />} favourite>
        Add
      </Button>
    )

    expect(screen.getByRole('button', { name: /Add/i })).toHaveStyle({
      background: '#FF4040',
      height: '4rem',
      padding: '0.8rem',
      'font-size': '1.6rem'
    })

    expect(screen.getByTestId('icon')).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render the add favorite button', () => {
    const { container } = renderWithTheme(
      <Button icon={<FavoriteBorder data-testid="icon" />} favourite={false}>
        Add to favourites
      </Button>
    )

    expect(screen.getByRole('button', { name: /Add/i })).toHaveStyle({
      height: '4rem',
      padding: '0.8rem',
      'font-size': '1.6rem'
    })

    expect(screen.getByTestId('icon')).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })
})
