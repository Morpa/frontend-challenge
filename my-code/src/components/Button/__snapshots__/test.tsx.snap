// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render the add favorite button 1`] = `
.c1 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  color: #FFF;
  border: 0;
  cursor: pointer;
  border-radius: 0.4rem;
  padding: 0.8rem;
  height: 4rem;
  font-size: 1.6rem;
  font-weight: 700;
  color: #7A8C99;
  background: none;
  border: 1px solid #7A8C99;
  border-radius: 0.4rem;
}

.c0 svg {
  width: 1.5rem;
}

.c0 svg + span {
  margin-left: 0.8rem;
}

.c0 svg {
  width: 1.5rem;
}

.c0 svg + span {
  margin-left: 0.8rem;
}

<button
  class="c0"
>
  <svg
    aria-hidden="true"
    class="c1"
    data-testid="icon"
    fill="currentColor"
    focusable="false"
    viewBox="0 0 24 24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M0 0h24v24H0V0z"
      fill="none"
    />
    <path
      d="M16.5 3c-1.74 0-3.41.81-4.5 2.09C10.91 3.81 9.24 3 7.5 3 4.42 3 2 5.42 2 8.5c0 3.78 3.4 6.86 8.55 11.54L12 21.35l1.45-1.32C18.6 15.36 22 12.28 22 8.5 22 5.42 19.58 3 16.5 3zm-4.4 15.55l-.1.1-.1-.1C7.14 14.24 4 11.39 4 8.5 4 6.5 5.5 5 7.5 5c1.54 0 3.04.99 3.57 2.36h1.87C13.46 5.99 14.96 5 16.5 5c2 0 3.5 1.5 3.5 3.5 0 2.89-3.14 5.74-7.9 10.05z"
    />
  </svg>
  <span>
    Add to favourites
  </span>
</button>
`;

exports[`<Button /> should render the favorited button 1`] = `
.c1 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  color: #FFF;
  border: 0;
  cursor: pointer;
  border-radius: 0.4rem;
  padding: 0.8rem;
  height: 4rem;
  font-size: 1.6rem;
  font-weight: 700;
  background: #FF4040;
}

.c0 svg {
  width: 1.5rem;
}

.c0 svg + span {
  margin-left: 0.8rem;
}

.c0 svg {
  color: #FFF;
  width: 1.5rem;
}

.c0 svg + span {
  margin-left: 0.8rem;
}

<button
  class="c0"
>
  <svg
    aria-hidden="true"
    class="c1"
    data-testid="icon"
    fill="currentColor"
    focusable="false"
    viewBox="0 0 24 24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M0 0h24v24H0V0z"
      fill="none"
    />
    <path
      d="M12 21.35l-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
    />
  </svg>
  <span>
    Add
  </span>
</button>
`;
